src/core/mac/mac.cpp

```diff
@@ -1620,83 +1620,77 @@ void Mac::HandleReceivedFrame(Frame *aFrame, otError aError)
-                      dstaddr.mShortAddress == mShortAddress), error = OT_ERROR_DESTINATION_ADDRESS_FILTERED);
+                     ((mRxOnWhenIdle && dstaddr.IsShortAddrBroadcast()) ||
+                      dstaddr.GetShort() == mShortAddress), error = OT_ERROR_DESTINATION_ADDRESS_FILTERED);
 
         // Allow  multicasts from neighbor routers if FFD
-        if (neighbor == NULL && dstaddr.mShortAddress == kShortAddrBroadcast &&
+        if (neighbor == NULL && dstaddr.IsShortAddrBroadcast() &&
             (GetNetif().GetMle().GetDeviceMode() & Mle::ModeTlv::kModeFFD))
         {
             neighbor = GetNetif().GetMle().GetRxOnlyNeighborRouter(srcaddr);
         }
 
         break;
 
-    case sizeof(ExtAddress):
+    case Address::kTypeExtended:
         aFrame->GetDstPanId(panid);
-        VerifyOrExit(panid == mPanId && dstaddr.mExtAddress == mExtAddress,
+        VerifyOrExit(panid == mPanId && dstaddr.GetExtended() == mExtAddress,
                      error = OT_ERROR_DESTINATION_ADDRESS_FILTERED);
         break;
     }
 
     // Source Address Filtering
-    switch (srcaddr.mLength)
+    switch (srcaddr.GetType())
     {
-    case 0:
+    case Address::kTypeEmpty:
         break;
 
-    case sizeof(ShortAddress):
-        otLogDebgMac(GetInstance(), "Received frame from short address 0x%04x", srcaddr.mShortAddress);
+    case Address::kTypeShort:
+        otLogDebgMac(GetInstance(), "Received frame from short address 0x%04x", srcaddr.GetShort());
 
         if (neighbor == NULL)
         {
             ExitNow(error = OT_ERROR_UNKNOWN_NEIGHBOR);
         }
 
-        srcaddr.mLength = sizeof(srcaddr.mExtAddress);
-        srcaddr.mExtAddress = neighbor->GetExtAddress();
-        break;
+        srcaddr.SetExtended(neighbor->GetExtAddress());
 
-    case sizeof(ExtAddress):
-        break;
+    // fall through
 
-    default:
-        ExitNow(error = OT_ERROR_INVALID_SOURCE_ADDRESS);
-    }
+    case Address::kTypeExtended:
 
-    // Duplicate Address Protection
-    if (srcaddr.mExtAddress == mExtAddress)
-    {
-        ExitNow(error = OT_ERROR_INVALID_SOURCE_ADDRESS);
-    }
+        // Duplicate Address Protection
+        if (srcaddr.GetExtended() == mExtAddress)
+        {
+            ExitNow(error = OT_ERROR_INVALID_SOURCE_ADDRESS);
+        }
 
 #if OPENTHREAD_ENABLE_MAC_FILTER
 
-    // Source filter Processing.
-    if (srcaddr.mLength != 0)
-    {
-        // check if filtered out by whitelist or blacklist.
-        SuccessOrExit(error = mFilter.Apply(srcaddr.mExtAddress, rssi));
+        // Source filter Processing. Check if filtered out by whitelist or blacklist.
+        SuccessOrExit(error = mFilter.Apply(srcaddr.GetExtended(), rssi));
 
         // override with the rssi in setting
         if (rssi != OT_MAC_FILTER_FIXED_RSS_DISABLED)
         {
             aFrame->mRssi = rssi;
         }
-    }
 
 #endif  // OPENTHREAD_ENABLE_MAC_FILTER
 
+        break;
+    }
+
```

Comment:
The diff may not be easy to read. The change is that the checks for duplicate address and mac filter are moved into the `switch` to avoid extra checks of address type....(from `case kTypeShort` we convert the address to extended and then fall through to `kTypeExtended` case where duplicate address and mac filter are checked).